File: cpp/cpp_100-999/919.cc
// C++
// leetcode 919
// https://leetcode.cn/problems/complete-binary-tree-inserter/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class CBTInserter {
public:
    int treeSize = 0;
    TreeNode *root;

    CBTInserter(TreeNode* r) {
        root = r;

        int res = 0;
        TreeNode *tmp = nullptr;

        queue<TreeNode*> q;
        q.push(r);
        while (!q.empty()) {
            int size = q.size();
            res += size;

            // check if it's the last level
            if (size * 2 -1 != res)
                break;

            // put next level into the queue
            while(size--) {
                tmp = q.front();
                q.pop();
                if (tmp -> left)
                    q.push(tmp -> left);
                else
                    break;
                if (tmp -> right)
                    q.push(tmp -> right);
                else
                    break;
            }

            while (size-- > 0)
                q.pop();
        }

        treeSize = res;
    }
    
    int insert(int val) {
        int cur = ++treeSize;
        int depth = 0;
        while(cur != 0) {
            cur >>= 1;
            depth += 1;
        }
        cur = treeSize;

        TreeNode *p = root;
        for (int i = depth - 2; i > 0; i--) {
            int u = 1 & (cur >> i);
            if (u == 0) {
                p = p -> left;
            } else {
                p = p -> right;
            }
        }
        // till now, p is the pointer of the father of new TreeNode

        // insert new TreeNode
        if ((cur & 1) == 0) {
            p -> left = new TreeNode(val);
        } else {
            p -> right = new TreeNode(val);
        }

        return p -> val;
    }
    
    TreeNode* get_root() {
        return root;
    }
};

/**
 * Your CBTInserter object will be instantiated and called as such:
 * CBTInserter* obj = new CBTInserter(root);
 * int param_1 = obj->insert(val);
 * TreeNode* param_2 = obj->get_root();
 */

File: Go/Go_100-999/919.go
// Go
// leetcode 919
// https://leetcode.cn/problems/complete-binary-tree-inserter/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type CBTInserter struct {
	TreeSize int
	Root *TreeNode
}


func Constructor(root *TreeNode) CBTInserter {
	res := 0
	q := []*TreeNode{root}
	for len(q) != 0 {
		size := len(q)
		res += size
		if size * 2 - 1 != res {
			break
		}

		for i := 0; i < size; i++ {
			if q[i].Left != nil {
				q = append(q, q[i].Left)
			} else {
				break
			}
			if q[i].Right != nil {
				q = append(q, q[i].Right)
			} else {
				break
			}
		}
		q = q[size : ]
	}

	return CBTInserter{res, root}
}


func (this *CBTInserter) Insert(val int) int {
	this.TreeSize++
	curIdx, depth := this.TreeSize, 0
	for curIdx > 0 {
		curIdx >>= 1
		depth++
	}
	curIdx = this.TreeSize
	p := this.Root

	// find father of new TreeNode
	for i := depth - 2; i > 0; i-- {
		if ((curIdx >> i) & 1) == 0 {
			p = p.Left
		} else {
			p = p.Right
		}
	}

	if curIdx & 1 == 0 {
		p.Left = &TreeNode{val, nil, nil}
	} else {
		p.Right = &TreeNode{val, nil, nil}
	}

	return p.Val
}


func (this *CBTInserter) Get_root() *TreeNode {
	return this.Root
}


/**
 * Your CBTInserter object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Insert(val);
 * param_2 := obj.Get_root();
 */